.buildkite/generate_pipeline.py

```diff
@@ -0,0 +1,240 @@
+
+The script will generate two pipelines:
+
+tests/smoke_tests
+├── test_*.py -> release pipeline
+├── test_required_before_merge.py -> pre-merge pipeline
+
+run `python .buildkite/generate_pipeline.py` to generate the pipeline for
+testing. The CI will run this script as a pre-step, and use the generated
+pipeline to run the tests.
+
+1. release pipeline, which runs all smoke tests by default, some function
+   support tests by multiple clouds, but we only generate one cloud per test
+   function to save cost.
+2. pre-merge pipeline, which generates all clouds supported by the test
+   function, author should specify which clouds to run by setting env in the
+   step.
+
+We only have credentials for aws/azure/gcp/kubernetes(CLOUD_QUEUE_MAP) now,
+smoke tests for those clouds are generated, other clouds are not supported
+yet, smoke tests for those clouds are not generated.
+"""
+
+import ast
+from collections import defaultdict
+import copy
+import os
+import random
+from typing import Any, Dict, List, Optional
+
+import yaml
+
+DEFAULT_CLOUDS_TO_RUN = ['aws', 'azure']
+
+ALL_CLOUDS_IN_SMOKE_TESTS = [
+    'aws', 'gcp', 'azure', 'lambda', 'cloudflare', 'ibm', 'scp', 'oci',
+    'kubernetes', 'vsphere', 'cudo', 'fluidstack', 'paperspace', 'runpod',
+    'lambda_cloud'
+]
+QUEUE_GENERIC_CLOUD = 'generic_cloud'
+QUEUE_GENERIC_CLOUD_SERVE = 'generic_cloud_serve'
+QUEUE_KUBERNETES = 'kubernetes'
+QUEUE_KUBERNETES_SERVE = 'kubernetes_serve'
+# Only aws, gcp, azure, and kubernetes are supported for now.
+# Other clouds do not have credentials.
+CLOUD_QUEUE_MAP = {
+    'aws': QUEUE_GENERIC_CLOUD,
+    'gcp': QUEUE_GENERIC_CLOUD,
+    'azure': QUEUE_GENERIC_CLOUD,
+    'kubernetes': QUEUE_KUBERNETES
+}
+# Serve tests runs long, and different test steps usually requires locks.
+# Its highly likely to fail if multiple serve tests are running concurrently.
+# So we use a different queue that runs only one concurrent test at a time.
+SERVE_CLOUD_QUEUE_MAP = {
+    'aws': QUEUE_GENERIC_CLOUD_SERVE,
+    'gcp': QUEUE_GENERIC_CLOUD_SERVE,
+    'azure': QUEUE_GENERIC_CLOUD_SERVE,
+    'kubernetes': QUEUE_KUBERNETES_SERVE
+}
+
+GENERATED_FILE_HEAD = ('# This is an auto-generated Buildkite pipeline by '
+                       '.buildkite/generate_pipeline.py, Please do not '
+                       'edit directly.\n')
+
+
+def _get_full_decorator_path(decorator: ast.AST) -> str:
+    """Recursively get the full path of a decorator."""
+    if isinstance(decorator, ast.Attribute):
+        return f'{_get_full_decorator_path(decorator.value)}.{decorator.attr}'
+    elif isinstance(decorator, ast.Name):
+        return decorator.id
+    elif isinstance(decorator, ast.Call):
+        return _get_full_decorator_path(decorator.func)
+    raise ValueError(f'Unknown decorator type: {type(decorator)}')
+
+
+def _extract_marked_tests(file_path: str) -> Dict[str, List[str]]:
+    """Extract test functions and filter clouds with pytest.mark
+    from a Python test file."""
```

Comment:
Why do we use this approach (parsing the ast and extracting which tests to run) rather than using pytest flags to filter based on marks? E.g. `pytest --serve`.
Could you add a comment explaining this?

Reply:
Update in the comment

```
    We separate each test_function_{cloud} into different pipeline steps
    to maximize the parallelism of the tests via the Buildkite CI job queue.
    This allows us to visualize the test results and rerun failures at the
    granularity of each test_function_{cloud}.

    If we make pytest --serve a job, it could contain dozens of test_functions
    and run for hours. This makes it hard to visualize the test results and
    rerun failures. Additionally, the parallelism would be controlled by pytest
    instead of the Buildkite job queue.
```